src/query/storage/m3/dynamic_cluster_test.go

```diff
@@ -0,0 +1,416 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/client"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/retention"
+	"github.com/m3db/m3/src/query/storage/m3/storagemetadata"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/instrument"
+	xtest "github.com/m3db/m3/src/x/test"
+	xwatch "github.com/m3db/m3/src/x/watch"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	defaultTestNs1ID         = ident.StringID("testns1")
+	defaultTestNs2ID         = ident.StringID("testns2")
+	defaultTestRetentionOpts = retention.NewOptions().
+					SetBufferFuture(10 * time.Minute).
+					SetBufferPast(10 * time.Minute).
+					SetBlockSize(2 * time.Hour).
+					SetRetentionPeriod(48 * time.Hour)
+	defaultTestNs2RetentionOpts = retention.NewOptions().
+					SetBufferFuture(10 * time.Minute).
+					SetBufferPast(10 * time.Minute).
+					SetBlockSize(4 * time.Hour).
+					SetRetentionPeriod(48 * time.Hour)
+	defaultTestAggregationOpts = namespace.NewAggregationOptions().
+					SetAggregations([]namespace.Aggregation{namespace.NewUnaggregatedAggregation()})
+	defaultTestNs2AggregationOpts = namespace.NewAggregationOptions().
+					SetAggregations([]namespace.Aggregation{namespace.NewAggregatedAggregation(
+			namespace.AggregatedAttributes{
+				Resolution:        1 * time.Minute,
+				DownsampleOptions: namespace.NewDownsampleOptions(true),
+			}),
+		})
+	defaultTestNs1Opts = namespace.NewOptions().SetRetentionOptions(defaultTestRetentionOpts).
+				SetAggregationOptions(defaultTestAggregationOpts)
+	defaultTestNs2Opts = namespace.NewOptions().SetRetentionOptions(defaultTestNs2RetentionOpts).
+				SetAggregationOptions(defaultTestNs2AggregationOpts)
+)
+
+func TestDynamicClustersInitialization(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	mockSession := client.NewMockSession(ctrl)
+
+	mapCh := make(nsMapCh, 10)
+	mapCh <- testNamespaceMap(t, []mapParams{
+		{nsID: defaultTestNs1ID, nsOpts: defaultTestNs1Opts},
+		{nsID: defaultTestNs2ID, nsOpts: defaultTestNs2Opts},
+	})
+	mockInitializer := newMockNsInitializer(t, ctrl, mapCh)
```

Comment:
nit: I think it's probably more straight forward here to use `namespace.NewMockInitializer` here?

Reply:
Renamed this newFakeNsInitializer since that's basically what it creates (and this is useful as I was able to inject an update channel we could use for cleaning up)